src/analysis/vector/abstracttool.cpp

```diff
@@ -0,0 +1,177 @@
+ *  abstracttool.h                                                      *
+ *  -------------------                                                    *
+ *  begin                : Jun 10, 2014                                    *
+ *  copyright            : (C) 2014 by Sandro Mani / Sourcepole AG         *
+ *  email                : smani@sourcepole.ch                             *
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QApplication>
+#include <QtConcurrentRun>
+#include <QtConcurrentMap>
+#include "abstracttool.h"
+#include "qgsfield.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgsvectordataprovider.h"
+#include "qgsvectorfilewriter.h"
+#include "qgsvectorlayer.h"
+#include "qgsgeometry.h"
+#include "qgslogger.h"
+#include "qgswkbtypes.h"
+#include "qgsgeos.h"
+
+namespace Vectoranalysis
+{
+
+  QString AbstractTool::errFeatureDoesNotExist = QApplication::translate( "AbstractTool", "The requested feature does not exist" );
+  QString AbstractTool::errFailedToFetchGeometry = QApplication::translate( "AbstractTool", "The feature geometry could not be fetched" );
+
+  void AbstractTool::ProcessFeatureWrapper::operator()( const Job *job )
+  {
+    try
+    {
+      instance->processFeature( job );
+    }
+    catch ( const std::exception &e )
+    {
+      instance->mExceptions.append( e.what() );
+      throw e;
+    }
+  }
+
+  AbstractTool::AbstractTool( QgsFeatureSink *output, QgsWkbTypes::Type outWkbType, double precision ): mOutput( output ), mOutWkbType( outWkbType ), mPrecision( precision )
+  {
+  }
+
+  AbstractTool::~AbstractTool()
+  {
+    qDeleteAll( mJobQueue );
+  }
+
+  QFuture<void> AbstractTool::init()
+  {
+    return QtConcurrent::run( this, &AbstractTool::prepare );
+  }
+
+  QFuture<void> AbstractTool::execute( int /*task*/ )
+  {
+    return QtConcurrent::map( mJobQueue, ProcessFeatureWrapper( this ) );
+  }
+
+  void AbstractTool::buildSpatialIndex( QgsSpatialIndex &index, QgsFeatureSource *layer ) const
+  {
+    QgsFeature currentFeature;
+    QgsFeatureRequest request;
+    request.setFlags( QgsFeatureRequest::SubsetOfAttributes );
+    QgsFeatureIterator it = layer->getFeatures();
+    while ( it.nextFeature( currentFeature ) )
+    {
+      index.addFeature( currentFeature );
```

Comment:
Would not be better to bulk-load features from the iterator or source into spatial index instead of adding them one by one? It should be faster.